File: cpp solution/valid palindrone.cpp
class Solution {
    bool is_valid(char ch){
        if ((ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || (ch >= '0' && ch <= '9')){
            return true;
        }
        return false;
    }
    char toLowercase(char ch){
        if ((ch >= 'a' && ch <= 'z') || (ch >= '0' && ch <= '9')){
            return ch;
        }
        else {
            char temp = ch - 'A' + 'a';
            return temp;
        }
    }
    bool is_palindrone(string S){
        int start = 0;
        int end = S.size() - 1;
        while (start < end){
            if (S[start] != S[end]){
                return false;
            }
            start++;
            end--;
        }
        return true;
    }
public:
    bool isPalindrome(string S) {
        string desired_str = "";
        for (auto ch: S){
            if (is_valid(ch)){
                desired_str += toLowercase(ch);
            }
        }
        bool result = is_palindrone(desired_str);
        return result;
    }
};

File: python solution/valid palindrone.py
class Solution:
    def check_palindrone(self, phrase):
        start = 0
        end = len(phrase) - 1
        while start < end:
            if phrase[start] == phrase[end]:
                return False
        return True
    def check_palindrone(self, phrase):
        start = 0
        end = len(phrase) - 1
        while start < end:
            if phrase[start] != phrase[end]:
                return False
            start += 1
            end -= 1
        return True
    def isPalindrome(self, s: str) -> bool:
        filtered_str = ""
        for ch in s:
            if ch.isalpha() or ch.isnumeric():
                filtered_str += ch.lower()
        return self.check_palindrone(filtered_str)
# one pass solution
class Solution:
    def isValid(self, ch):
        return ch.isalpha() or ch.isnumeric()
    
    def isPalindrome(self, s: str) -> bool:
        start = 0
        end = len(s) - 1
        while start < end:
            if not self.isValid(s[start]):
                start += 1
                continue
            if not self.isValid(s[end]):
                end -= 1
                continue
            if s[start].lower() != s[end].lower():
                return False
            start += 1
            end -= 1
        return True
    # changes made here
